src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureKeyVaultCredentialRetriever.java

```diff
@@ -57,4 +62,55 @@ public static KeyVaultCredentials getCredentialById(String credentialID, Run<?,
 
         return credential;
     }
+
+    public static KeyVaultCredentials getCredentialById(String credentialID) {
+        SystemCredentialsProvider systemCredentialsProvider = SystemCredentialsProvider.getInstance();
+        List<AzureImdsCredentials> azureImdsCredentials =
+                DomainCredentials.getCredentials(systemCredentialsProvider.getDomainCredentialsMap(),
+                        AzureImdsCredentials.class,
+                        Collections.emptyList(),
+                        CredentialsMatchers.withId(credentialID));
+
+        if (!azureImdsCredentials.isEmpty()) {
+            return new AzureKeyVaultImdsCredential();
+        }
+
+        AzureKeyVaultCredential credential = null;
+        List<StandardUsernamePasswordCredentials> usernamePasswordCredentials =
+                DomainCredentials.getCredentials(systemCredentialsProvider.getDomainCredentialsMap(),
+                        StandardUsernamePasswordCredentials.class,
+                        Collections.emptyList(),
+                        CredentialsMatchers.withId(credentialID));
+        if (!usernamePasswordCredentials.isEmpty()) {
+            LOGGER.log(Level.FINE, format("Fetched %s as StandardUsernamePasswordCredentials", credentialID));
+            StandardUsernamePasswordCredentials usernamePasswordCredential = usernamePasswordCredentials.get(0);
+            credential = new AzureKeyVaultCredential(
+                    usernamePasswordCredential.getUsername(),
+                    usernamePasswordCredential.getPassword()
+            );
+        }
+
+        List<AzureCredentials> azureCredentials =
```

Comment:
I can't seem to get this to work,

I have a service principal configured, "Microsoft Azure Service Principal"

the credentials get's found by ID, but then it tries to fetch the credential from key vault rather than Jenkins' credential store

I don't think this credential provider is playing nice with:

1. other credential providers
2. I can't see any code filtering for only string / pfx credentials, (due to azure keyvault only taking one value I can't see how we can handle non string credentials (unless they get encoded in json i guess).

Reply:
http://localhost:8088/credentials/store/system/domain/_/credential/jenkinsServicePrincipal/update
![image](https://user-images.githubusercontent.com/21194782/72785269-bfa46c80-3c22-11ea-8144-b77b4a831dfd.png)

Reply:
Hmm other credentials seem to work that I have installed, its just the Azure one that is trying to contact key vault

Reply:
Hmm 2. manually updating the SP seems to have fixed it,
note this was created via jcasc

Reply:
ah turns out it was my mistake, I didn't have the cred in my environment, fixed now